File: Rapi-frontend-master/src/services/rapi-backend/interfaceinfoController.ts
// @ts-ignore
/* eslint-disable */
import { request } from '@umijs/max';

/** addInterfaceinfo POST /api/interfaceinfo/add */
export async function addInterfaceinfoUsingPost(
  body: API.InterfaceinfoAddRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseLong_>('/api/interfaceinfo/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** deleteInterfaceinfo POST /api/interfaceinfo/delete */
export async function deleteInterfaceinfoUsingPost(
  body: API.DeleteRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/interfaceinfo/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** getInterfaceinfoVOById GET /api/interfaceinfo/get/vo */
export async function getInterfaceinfoVoByIdUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getInterfaceinfoVOByIdUsingGETParams,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseInterfaceinfoVO_>('/api/interfaceinfo/get/vo', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** invokeInterfaceinfo POST /api/interfaceinfo/invoke */
export async function invokeInterfaceinfoUsingPost(
  body: API.InterfaceinfoInvokeRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseObject_>('/api/interfaceinfo/invoke', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** listInterfaceinfoByPage POST /api/interfaceinfo/list/page */
export async function listInterfaceinfoByPageUsingPost(
  body: API.InterfaceinfoQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageInterfaceInfo_>('/api/interfaceinfo/list/page', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** listInterfaceinfoVOByPage POST /api/interfaceinfo/list/page/vo */
export async function listInterfaceinfoVoByPageUsingPost(
  body: API.InterfaceinfoQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageInterfaceinfoVO_>('/api/interfaceinfo/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** listMyInterfaceinfoVOByPage POST /api/interfaceinfo/my/list/page/vo */
export async function listMyInterfaceinfoVoByPageUsingPost(
  body: API.InterfaceinfoQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageInterfaceinfoVO_>('/api/interfaceinfo/my/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** offlineInterfaceinfo POST /api/interfaceinfo/offline */
export async function offlineInterfaceinfoUsingPost(
  body: API.IdRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/interfaceinfo/offline', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** onlineInterfaceinfo POST /api/interfaceinfo/online */
export async function onlineInterfaceinfoUsingPost(
  body: API.IdRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/interfaceinfo/online', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** updateInterfaceinfo POST /api/interfaceinfo/update */
export async function updateInterfaceinfoUsingPost(
  body: API.InterfaceinfoUpdateRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/interfaceinfo/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

File: Rapi-backend/src/main/java/com/yupi/springbootinit/controller/InterfaceinfoController.java
package com.yupi.springbootinit.controller;


import cn.hutool.core.date.DateTime;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.google.gson.Gson;
import com.yupi.rapiclientsdk.client.RapiClient;
import com.yupi.rapicommon.model.entity.InterfaceInfo;
import com.yupi.rapicommon.model.entity.User;
import com.yupi.springbootinit.annotation.AuthCheck;
import com.yupi.springbootinit.common.*;
import com.yupi.springbootinit.constant.UserConstant;
import com.yupi.springbootinit.exception.BusinessException;
import com.yupi.springbootinit.exception.ThrowUtils;
import com.yupi.springbootinit.model.dto.interfaceinfo.InterfaceinfoAddRequest;
import com.yupi.springbootinit.model.dto.interfaceinfo.InterfaceinfoInvokeRequest;
import com.yupi.springbootinit.model.dto.interfaceinfo.InterfaceinfoQueryRequest;
import com.yupi.springbootinit.model.dto.interfaceinfo.InterfaceinfoUpdateRequest;
import com.yupi.springbootinit.model.enums.InterfaceinfoStatusEnums;
import com.yupi.springbootinit.model.vo.InterfaceinfoVO;
import com.yupi.springbootinit.service.InterfaceInfoService;
import com.yupi.springbootinit.service.UserService;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.BeanUtils;
import org.springframework.web.bind.annotation.*;
import javax.annotation.Resource;
import javax.servlet.http.HttpServletRequest;

/**
 * 接口管理
 *
 */
@RestController
@RequestMapping("/interfaceinfo")
@Slf4j
public class InterfaceinfoController {

    @Resource
    private InterfaceInfoService interfaceinfoService;

    @Resource
    private UserService userService;

    @Resource
    private RapiClient rapiClient;

    // region 增删改查

    /**
     * 创建
     *
     * @param interfaceinfoAddRequest
     * @param request
     * @return
     */
    @PostMapping("/add")
    public BaseResponse<Long> addInterfaceinfo(@RequestBody InterfaceinfoAddRequest interfaceinfoAddRequest, HttpServletRequest request) {
        if (interfaceinfoAddRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        InterfaceInfo interfaceinfo = new InterfaceInfo();
        BeanUtils.copyProperties(interfaceinfoAddRequest, interfaceinfo);
        interfaceinfo.setCreateTime(DateTime.now());
        interfaceinfo.setUpdateTime(DateTime.now());

        // 参数校验
        interfaceinfoService.validInterfaceinfo(interfaceinfo, true);
        User loginUser = userService.getLoginUser(request);
        interfaceinfo.setUserId(loginUser.getId());

        boolean result = interfaceinfoService.save(interfaceinfo);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        long newInterfaceinfoId = interfaceinfo.getId();
        return ResultUtils.success(newInterfaceinfoId);
    }

    /**
     * 删除
     *
     * @param deleteRequest
     * @param request
     * @return
     */
    @PostMapping("/delete")
    public BaseResponse<Boolean> deleteInterfaceinfo(@RequestBody DeleteRequest deleteRequest, HttpServletRequest request) {
        if (deleteRequest == null || deleteRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User user = userService.getLoginUser(request);
        long id = deleteRequest.getId();
        // 判断是否存在
        InterfaceInfo oldInterfaceinfo = interfaceinfoService.getById(id);
        ThrowUtils.throwIf(oldInterfaceinfo == null, ErrorCode.NOT_FOUND_ERROR);
        // 仅本人或管理员可删除
        if (!oldInterfaceinfo.getUserId().equals(user.getId()) && !userService.isAdmin(request)) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR);
        }
        boolean b = interfaceinfoService.removeById(id);
        return ResultUtils.success(b);
    }

    /**
     * 更新（仅管理员）
     *
     * @param interfaceinfoUpdateRequest
     * @return
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> updateInterfaceinfo(@RequestBody InterfaceinfoUpdateRequest interfaceinfoUpdateRequest) {
        if (interfaceinfoUpdateRequest == null || interfaceinfoUpdateRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        InterfaceInfo interfaceinfo = new InterfaceInfo();
        BeanUtils.copyProperties(interfaceinfoUpdateRequest, interfaceinfo);
        interfaceinfo.setUpdateTime(DateTime.now());

        // 参数校验
        interfaceinfoService.validInterfaceinfo(interfaceinfo, false);
        long id = interfaceinfoUpdateRequest.getId();
        // 判断是否存在
        InterfaceInfo oldInterfaceinfo = interfaceinfoService.getById(id);
        ThrowUtils.throwIf(oldInterfaceinfo == null, ErrorCode.NOT_FOUND_ERROR);
        boolean result = interfaceinfoService.updateById(interfaceinfo);
        return ResultUtils.success(result);
    }

    /**
     * 根据 id 获取
     *
     * @param id
     * @return
     */
    @GetMapping("/get/vo")
    public BaseResponse<InterfaceinfoVO> getInterfaceinfoVOById(long id, HttpServletRequest request) {
        if (id <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        InterfaceInfo interfaceinfo = interfaceinfoService.getById(id);
        if (interfaceinfo == null) {
            throw new BusinessException(ErrorCode.NOT_FOUND_ERROR);
        }
        return ResultUtils.success(interfaceinfoService.getInterfaceinfoVO(interfaceinfo, request));
    }

    /**
     * 分页获取列表（仅管理员）
     *
     * @param interfaceinfoQueryRequest
     * @return
     */
    @PostMapping("/list/page")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Page<InterfaceInfo>> listInterfaceinfoByPage(@RequestBody InterfaceinfoQueryRequest interfaceinfoQueryRequest) {
        long current = interfaceinfoQueryRequest.getCurrent();
        long size = interfaceinfoQueryRequest.getPageSize();
        if(size >50 ){
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }

        Page<InterfaceInfo> interfaceinfoPage = interfaceinfoService.page(new Page<>(current, size),
                interfaceinfoService.getQueryWrapper(interfaceinfoQueryRequest));
        return ResultUtils.success(interfaceinfoPage);
    }

    /**
     * 分页获取列表（封装类）
     *
     * @param interfaceinfoQueryRequest
     * @param request
     * @return
     */
    @PostMapping("/list/page/vo")
    public BaseResponse<Page<InterfaceinfoVO>> listInterfaceinfoVOByPage(@RequestBody InterfaceinfoQueryRequest interfaceinfoQueryRequest,
            HttpServletRequest request) {
        long current = interfaceinfoQueryRequest.getCurrent();
        long size = interfaceinfoQueryRequest.getPageSize();
        // 限制爬虫
        ThrowUtils.throwIf(size > 20, ErrorCode.PARAMS_ERROR);
        Page<InterfaceInfo> interfaceinfoPage = interfaceinfoService.page(new Page<>(current, size),
                interfaceinfoService.getQueryWrapper(interfaceinfoQueryRequest));
        return ResultUtils.success(interfaceinfoService.getInterfaceinfoVOPage(interfaceinfoPage, request));
    }

    /**
     * 分页获取当前用户创建的资源列表
     *
     * @param interfaceinfoQueryRequest
     * @param request
     * @return
     */
    @PostMapping("/my/list/page/vo")
    public BaseResponse<Page<InterfaceinfoVO>> listMyInterfaceinfoVOByPage(@RequestBody InterfaceinfoQueryRequest interfaceinfoQueryRequest,
            HttpServletRequest request) {
        if (interfaceinfoQueryRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User loginUser = userService.getLoginUser(request);
        interfaceinfoQueryRequest.setUserId(loginUser.getId());
        long current = interfaceinfoQueryRequest.getCurrent();
        long size = interfaceinfoQueryRequest.getPageSize();
        // 限制爬虫
        ThrowUtils.throwIf(size > 20, ErrorCode.PARAMS_ERROR);
        Page<InterfaceInfo> interfaceinfoPage = interfaceinfoService.page(new Page<>(current, size),
                interfaceinfoService.getQueryWrapper(interfaceinfoQueryRequest));
        return ResultUtils.success(interfaceinfoService.getInterfaceinfoVOPage(interfaceinfoPage, request));
    }


    /**
     * 上线接口（仅管理员）
     *
     * @param idRequest
     * @return
     */
    @PostMapping("/online")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> onlineInterfaceinfo(@RequestBody IdRequest idRequest) {
        if (idRequest == null || idRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        // 判断是否存在
        Long id = idRequest.getId();
        InterfaceInfo oldInterfaceinfo = interfaceinfoService.getById(id);
        ThrowUtils.throwIf(oldInterfaceinfo == null, ErrorCode.NOT_FOUND_ERROR);

        //先调试判断一下该接口是否可以调用，可以正常调用的才能上线
        com.yupi.rapiclientsdk.modal.User user = new com.yupi.rapiclientsdk.modal.User();
        user.setUsername("ruiTest");
        String re = rapiClient.getUserNameByPost(user);

        if(!StringUtils.isNotBlank(re)){
            throw new BusinessException(ErrorCode.SYSTEM_ERROR,"接口调用失败");
        }

        //更新接口状态
        oldInterfaceinfo.setStatus(InterfaceinfoStatusEnums.ONLINE.getValue());
        boolean result = interfaceinfoService.updateById(oldInterfaceinfo);

        return ResultUtils.success(result);
    }


    /**
     * 下线接口（仅管理员）
     *
     * @param idRequest
     * @return
     */
    @PostMapping("/offline")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> offlineInterfaceinfo(@RequestBody IdRequest idRequest) {
        if (idRequest == null || idRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        // 判断是否存在
        Long id = idRequest.getId();
        InterfaceInfo oldInterfaceinfo = interfaceinfoService.getById(id);
        ThrowUtils.throwIf(oldInterfaceinfo == null, ErrorCode.NOT_FOUND_ERROR);

        //更新接口状态
        oldInterfaceinfo.setStatus(InterfaceinfoStatusEnums.OFFLINE.getValue());
        boolean result = interfaceinfoService.updateById(oldInterfaceinfo);

        return ResultUtils.success(result);
    }


    /**
     * 调用接口
     *
     * @param interfaceinfoInvokeRequest
     * @return
     */
    @PostMapping("/invoke")
//    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Object> invokeInterfaceinfo(@RequestBody InterfaceinfoInvokeRequest interfaceinfoInvokeRequest,HttpServletRequest httpServletRequest) {
        if (interfaceinfoInvokeRequest == null || interfaceinfoInvokeRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        // 判断是否存在
        Long id = interfaceinfoInvokeRequest.getId();
        InterfaceInfo oldInterfaceinfo = interfaceinfoService.getById(id);
        ThrowUtils.throwIf(oldInterfaceinfo == null, ErrorCode.NOT_FOUND_ERROR);
        if(oldInterfaceinfo.getStatus() == InterfaceinfoStatusEnums.OFFLINE.getValue()){
            throw new BusinessException(ErrorCode.PARAMS_ERROR,"接口已关闭");
        }

        String userRequestParams = interfaceinfoInvokeRequest.getUserRequestParams();
        User loginUser = userService.getLoginUser(httpServletRequest);
        String accessKey = loginUser.getAccessKey();
        String secretKey = loginUser.getSecretKey();

        RapiClient templeClient = new RapiClient(accessKey,secretKey);
        Gson gson = new Gson();
        com.yupi.rapiclientsdk.modal.User user = gson.fromJson(userRequestParams, com.yupi.rapiclientsdk.modal.User.class);
        String userNameByPost = templeClient.getUserNameByPost(user);


        return ResultUtils.success(userNameByPost);
    }

}
